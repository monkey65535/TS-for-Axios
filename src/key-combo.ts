/*
  Copyright 2017 Alex Jover Morales (alexjovermorales@gmail.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { keyContainer } from './key-container'

/**
 * Simple KeyCombo class
 * @class KeyCombo
 */
export class KeyCombo {
  public keys: Set<number|number[]>
  private comboStr: string

  /**
   * Creates an instance of KeyCombo.
   * @param {array} keys
   */
  constructor (comboStr: string) {
    this.comboStr = comboStr
    this.keys = new Set(this.parse(comboStr))
  }

  /**
   * Parses the string and gets the code or codes of it
   * @param {string} comboStr
   * @returns {number[]}
   */
  private parse (comboStr: string): (number|number[])[] {
    let comboArr: string[] = comboStr.trim().replace(/\s+/g, ' ').toLowerCase().split(' ')
    let comboArrParsed: (number|number[])[] = comboArr.map(combo => keyContainer.getValue(combo))
    return comboArrParsed
  }
}
