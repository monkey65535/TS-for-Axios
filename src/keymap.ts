
/**
 * Mapping of keyCodes to properties.
 *
 * The idea of the mapping
 */
export default {
  shift: 16,
  ctrl: 17,
  backslash: 220
}
