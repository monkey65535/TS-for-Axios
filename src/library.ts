// Import here Polyfills if needed. Recommended core-js (npm i -D core-js)
  // import "core-js/fn/array.find"
  // ...

const a = new Set<number>()

export default class DummyClass {

}
