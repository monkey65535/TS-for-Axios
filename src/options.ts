export interface IOptions {
  debug?: boolean
  stopPropagation?: boolean
}

export class Options implements IOptions {
  public debug: boolean = false
  public stopPropagation: boolean = false

  constructor(obj?: IOptions) {
    if (obj && Object.keys(obj).some(key => typeof this[key] === 'undefined')) {
      throw new Error('Some of the options are not correct. Checkout the docs at https://github.com/coosto/ShortcutJS for more info')
    } else if (obj) {
      Object.assign(this, obj)
    }
  }
}
