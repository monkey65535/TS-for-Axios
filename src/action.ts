import { KeyCombo } from './key-combo'

/**
 * Action class to use for the ShortcutJS
 * @export
 * @class Action
 */
export class Action {
  public name: string
  public keyCombo: KeyCombo
  public callbacks: Set<Function>
  /**
   * Creates an instance of Action.
   * @param {sring} name
   * @param {KeyCombo} keyCombo
   */
  constructor (name: string, keyCombo: KeyCombo) {
    if (!(keyCombo instanceof KeyCombo)) {
      throw new Error('The second parameter (keyCombo) must be an instance of KeyCombo')
    }

    this.name = name
    this.keyCombo = keyCombo
    this.callbacks = new Set()
  }
}