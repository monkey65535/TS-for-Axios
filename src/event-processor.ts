import { Action } from './action'
import { KeyCombo } from './key-combo'
import { logger, compareSets } from './utils'

export class EventProcessor {
  public currentCombo: KeyCombo
  public lastEvent: KeyboardEvent

  constructor () {
    this.currentCombo = new KeyCombo()
  }

  public reset () {
    this.currentCombo = new KeyCombo()
  }

  public processEvent (ev: KeyboardEvent, actions: Map<string, Action>, debug: boolean) {
    const wasAppended = this.currentCombo.addEvent(ev)

    // Avoid repeated events
    if (!wasAppended) {
      return false
    }

    if (debug) {
      this.printDebugKeyPressed(ev)
    }

    this.processActionCombos(ev, actions, debug)
  }

  public cleanCombo (debug) {
    this.reset()

    if (debug) {
      logger.log('ShortcutJS: Cleaned keyMap')
    }
  }

  /**
   * Search for the right action, given a keyCombo, and execute its callbacks
   */
  public processActionCombos (ev: KeyboardEvent, actions: Map<string, Action>, debug: boolean) {
    for (let action of actions.values()) {
      if (this.matchesComboAction(action)) {
        if (debug) {
          this.printDebugActionFound(action)
        }

        for (let cb of action.callbacks) {
          cb(ev)
        }
        // Don't continue after finding it
        return false
      }
    }
  }

  /**
   * Checks whether the que matches a particular action
   */
  private matchesComboAction (action: Action) {
    return KeyCombo.isEqual(this.currentCombo, action.keyCombo)
  }

  private printDebugKeyPressed (ev: KeyboardEvent) {
    logger.group('ShortcutJS: KeyPressed')
    logger.log('Key: ', ev.keyCode)
    logger.group('Current combo:')
    logger.log('Keys: ', [...this.currentCombo.keys])
    logger.log('State keys: ', this.currentCombo.stateKeys)
    logger.groupEnd()
    logger.groupEnd()
  }

  private printDebugActionFound (action: Action) {
    logger.group('%cShortcutJS: Action Matched', 'color: green')
    logger.log('Action: ', action.name)
    logger.group('Current combo:')
    logger.log('Keys: ', [...this.currentCombo.keys])
    logger.log('State keys: ', this.currentCombo.stateKeys)
    logger.groupEnd()
    logger.log(`${action.callbacks.size} callbacks found`)
    logger.groupEnd()
  }
}
