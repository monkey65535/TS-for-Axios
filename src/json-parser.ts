import { ShortcutJS } from './index'
import { Action } from './action'
import { KeyCombo } from './key-combo'

export class JsonActionCombo {
  public combo
  public action

  constructor(obj) {
    if (!obj || !obj.combo || !obj.action) {
      throw new Error(`
        The json provided must be an array of { combo, action } object. Example
          [
            { action: 'openWindow', combo: 'ctrl a' }
          ]
      `)
    }

    this.combo = obj.combo
    this.action = obj.action
  }
}

export class JsonParser {
  static parse(shortcutJS: ShortcutJS, json: any) {
    if (!Array.isArray(json)) {
      throw new Error('The json provided must be an array')
    }

    json.forEach(obj => {
      const jsonActionCombo = new JsonActionCombo(obj)
      const keyCombo = new KeyCombo(jsonActionCombo.combo)
      const action = new Action(jsonActionCombo.action, keyCombo)
      shortcutJS.addAction(action)
    })
  }
}