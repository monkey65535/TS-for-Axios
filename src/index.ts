/*
  Copyright 2017 Alex Jover Morales (alexjovermorales@gmail.com)

  Licensed under the Apache License, Version 2.0 (the "License");
  you may not use this file except in compliance with the License.
  You may obtain a copy of the License at

      http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
 */

import { keyContainer } from './key-container'
import { JsonParser } from './json-parser'
import { KeyCombo } from './key-combo'
import { Action } from './action'

/**
 * First Version of a ShortcutJS.
 *
 * Ideally, ShortcutJS could hold the keydown and keyup events, and execute the actions according
 * to the key combos.
 *
 * WARNING: careful when adding combos with CTRL key on it. If you happen to trigger a by-default
 * browser shortcut (like CTRL-T) the keyup event will not be performed
 *
 * @todo Improvements:
 *  - Avoid repetition of mouseDownEvents if the last key is the same
 *  - Performance of processActionCombos
 *
 * @class ShortcutJS
 */
export class ShortcutJS {
  public actions: Map<string, Action>
  public keyMap: Map<string, boolean>
  public options: any

  // ********* PRIVATE API (only internal use) **********
  /**
   * Private constructor
   *
   * @memberOf ShortcutJS
   */
  constructor () {
    // With Maps we avoid events duplication, achieve immutability and performance
    this.actions = new Map()
    this.keyMap = new Map()
    this.options = {
      debug: false
    }
  }

  public init (options) {
    // @todo do some checking here
    this.options = Object.assign({}, this.options, options)


    window.addEventListener('keydown', this.processEvent.bind(this))
    window.addEventListener('keyup', this.removeAllKeys.bind(this))
  }

  public loadFromJson(json, options = null) {
    if (options) {
      this.init(options)
    }

    JsonParser.parse(this, json)
  }

  public addAction (action) {
    if (!(action instanceof Action)) {
      throw new Error('You must pass an Action instance object')
    }

    this.actions.set(action.name, action)
  }

  public subscribe (actionName, cb) {
    if (this.actions.has(actionName)) {
      const action = this.actions.get(actionName)
      action.callbacks.add(cb)
    } else {
      throw new Error(`Action ${actionName} does not exists`)
    }
  }

  public unsubscribe (actionName, cb = null) {
    if (this.actions.has(actionName)) {
      const action = this.actions.get(actionName)

      if (cb) {
        action.callbacks.delete(cb)
      } else {
        action.callbacks = new Set()
      }
    } else {
      throw new Error(`Action ${actionName} does not exists`)
    }
  }

  private addEventToQueue (ev) {
    if (this.keyMap.has(ev.keyCode)) {
      return false
    } else {
      this.keyMap.set(ev.keyCode, true)
      return true
    }
  }

  /**
   * Cleans the event Queue
   */
  private removeAllKeys (ev) {
    this.keyMap.clear()
    if (this.options.debug) {
      console.log('ShortcutJS: Cleaned keyMap')
    }
  }

  /**
   * Checks whether the que matches a particular action
   * @todo Performance improvement
   *   - Idea 1: Using binary operators for comparison
   *   - Idea 2: Generating uuid's per KeyCombo
   * @param {any} action
   */
  private isQueueInAction (action) {
    const diff = new Set([...action.keyCombo.keys.keys()].filter(key => !this.keyMap.has(key)))
    return !diff.size // return boolean from size
  }

  /**
   * Search for the right action, given a keyCombo, and execute its callbacks
   */
  private processActionCombos () {
    for (let action of this.actions.values()) {
      if (this.isQueueInAction(action)) {
        this.printDebugActionFound(action)
        for (let cb of action.callbacks) {
          cb()
        }
        // Don't continue after finding it
        return false
      }
    }
  }

  private printDebugKeyPressed (ev: KeyboardEvent) {
    console.group('ShortcutJS: KeyPressed')
    console.log('Key: ', ev.keyCode)
    console.log('Current keyMap: ', [...this.keyMap.keys()])
    console.groupEnd()
  }

  private printDebugActionFound (action: Action) {
    console.group('ShortcutJS: Action Matched')
    console.log('Action: ', action.name)
    console.log('Current keyMap: ', [...this.keyMap.keys()])
    console.log(`${action.callbacks.size} callbacks found`)
    console.groupEnd()
  }

  private processEvent (ev) {
    const wasAppended = this.addEventToQueue(ev)

    if (this.options.debug) {
      this.printDebugKeyPressed(ev)
    }

    if (wasAppended) {
      this.processActionCombos()
    }
  }
}

export const shortcutJS = new ShortcutJS() // Enforce singleton
export { KeyCombo } from './key-combo'
export { Action } from './action'
