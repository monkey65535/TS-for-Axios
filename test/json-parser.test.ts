import { JsonActionCombo, JsonParser } from 'src/json-parser'
import { shortcutJS } from 'src/index'

describe('JsonParser', () => {
  it('throws an error if json is not array', () => {
    expect(() => JsonParser.parse(shortcutJS, {})).toThrowErrorMatchingSnapshot()
  })
})
