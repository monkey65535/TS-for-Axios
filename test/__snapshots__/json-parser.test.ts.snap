exports[`JsonParser throws an error if json is not array 1`] = `"The json provided must be an array"`;
